File: project_euler/Problem066/Problem066.java
package project_euler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Problem066 {
    public void solution() {
        int target = 1000;
        List<Integer> factors = new ArrayList<>();
        Decimal previous = new Decimal(100, 0);
        Decimal current = new Decimal(100, 0);
        Decimal auxiliary = new Decimal(100, 0);

        int result = 0;
        Decimal pivot = new Decimal(100, 0);
        for (int i = 5; i <= target; ++i) {
            if (getSolution(i, factors, previous, current, auxiliary) && isGreater(current, pivot)) {
                swapDecimal(current, pivot);
                result = i;
            }
        }
        System.out.println(result);
    }

    private boolean getSolution(int value, List<Integer> factors, Decimal previous, Decimal current, Decimal auxiliary) {
        factors.clear();
        int start = getPeriod(value, factors);
        if (0 == start) {
            return false;
        }

        setValue(0, previous);
        setValue(1, current);
        getConvergents(previous, current, factors);
        if (0 != (factors.size() & 1)) {
            return true;
        }

        setValue(1, previous);
        setValue(start, auxiliary);
        getConvergents(previous, auxiliary, factors);
        swapDecimal(previous, current);
        getProduct(previous, auxiliary, current);
        getDouble(current);
        return true;
    }

    private boolean isGreater(Decimal current, Decimal pivot) {
        if (current.length != pivot.length) {
            return current.length > pivot.length;
        }

        for (int i = current.length - 1; i >= 0; --i) {
            if (current.value[i] != pivot.value[i]) {
                return current.value[i] > pivot.value[i];
            }
        }
        return false;
    }

    private void swapDecimal(Decimal previous, Decimal current) {
        int length = Math.max(previous.length, current.length);
        for (int i = 0; i < length; ++i) {
            previous.value[i] ^= current.value[i];
            current.value[i] ^= previous.value[i];
            previous.value[i] ^= current.value[i];
        }
        previous.length ^= current.length;
        current.length ^= previous.length;
        previous.length ^= current.length;
    }

    private int getPeriod(int value, List<Integer> factors) {
        if (Math.abs(Math.floor(Math.sqrt(value) + 0.5) - Math.sqrt(value)) < 0.00000001) {
            return 0;
        }

        int initial = (int) Math.sqrt(value);
        int pivot = initial << 1;
        int current = initial;
        int offset = 0;
        int factor = 1;
        while (current != pivot) {
            offset = current * factor - offset;
            factor = (value - offset * offset) / factor;
            current = (initial + offset) / factor;
            factors.add(current);
        }
        factors.remove(factors.size() - 1);
        return initial;
    }

    private void setValue(int value, Decimal decimal) {
        Arrays.fill(decimal.value, 0);
        decimal.value[0] = value;
        decimal.length = 1;
    }

    private void getConvergents(Decimal previous, Decimal current, List<Integer> factors) {
        for (int factor : factors) {
            getSummation(current, factor, previous);
            swapDecimal(previous, current);
        }
    }

    private void getProduct(Decimal multiplicand, Decimal multiplier, Decimal product) {
        for (int i = 0; i < multiplicand.length; ++i) {
            for (int j = 0; j < multiplier.length; ++j) {
                long value = (long) multiplicand.value[i] * multiplier.value[j];
                int index = i + j;
                product.value[index] += (int) (value % (int) 1e5);
                product.value[index + 1] += (int) (value / (int) 1e5);
            }
        }
        product.length = multiplicand.length + multiplier.length - 1;
        carry(product);
    }

    private void getDouble(Decimal decimal) {
        for (int i = 0; i < decimal.length; ++i) {
            decimal.value[i] <<= 1;
        }
        carry(decimal);
    }

    private void getSummation(Decimal decimal, int factor, Decimal summation) {
        for (int i = 0; i < decimal.length; ++i) {
            summation.value[i] += decimal.value[i] * factor;
        }
        carry(summation);
    }

    private void carry(Decimal decimal) {
        for (int i = 0; i < decimal.length; ++i) {
            if (decimal.value[i] >= (int) 1e5) {
                decimal.value[i + 1] += decimal.value[i] / (int) 1e5;
                decimal.value[i] %= (int) 1e5;
            }
        }
        if (0 != decimal.value[decimal.length]) {
            ++decimal.length;
        }
    }

    private static class Decimal {
        public int[] value;
        public int length;

        public Decimal(int capacity, int length) {
            this.value = new int[capacity];
            Arrays.fill(this.value, 0);
            this.length = length;
        }
    }
}

File: project_euler/Problem066/Problem066.js
function Problem066() {
    let target = 1000
    let factors = []
    let previous = new Decimal(100, 0)
    let current = new Decimal(100, 0)
    let auxiliary = new Decimal(100, 0)

    let result = 0
    let pivot = new Decimal(100, 0)
    for (let i = 5; i <= target; ++i) {
        if (getSolution(i, factors, previous, current, auxiliary) && isGreater(current, pivot)) {
            swapDecimal(current, pivot)
            result = i
        }
    }
    console.log(result)
}

function getSolution(value, factors, previous, current, auxiliary) {
    factors.length = 0
    let start = getPeriod(value, factors)
    if (!start) {
        return false;
    }

    setValue(0, previous)
    setValue(1, current)
    getConvergents(previous, current, factors)
    if (0 !== (factors.length & 1)) {
        return true;
    }

    setValue(1, previous)
    setValue(start, auxiliary)
    getConvergents(previous, auxiliary, factors)
    swapDecimal(previous, current)
    getProduct(previous, auxiliary, current)
    getDouble(current)
    return true
}

function isGreater(current, pivot) {
    if (current.length !== pivot.length) {
        return current.length > pivot.length
    }

    for (let i = current.length - 1; i >= 0; --i) {
        if (current.value[i] !== pivot.value[i]) {
            return current.value[i] > pivot.value[i]
        }
    }
    return false
}

function swapDecimal(previous, current) {
    let length = Math.max(previous.length, current.length)
    for (let i = 0; i < length; ++i) {
        previous.value[i] ^= current.value[i]
        current.value[i] ^= previous.value[i]
        previous.value[i] ^= current.value[i]
    }
    previous.length ^= current.length
    current.length ^= previous.length
    previous.length ^= current.length
}

function getPeriod(value, factors) {
    let auxiliary = Math.sqrt(value)
    if (Math.abs(Math.floor(auxiliary + 0.5) - auxiliary) < 0.00000001) {
        return 0
    }

    let initial = Math.floor(auxiliary)
    let pivot = initial << 1
    let current = initial
    let offset = 0
    let factor = 1
    while (current !== pivot) {
        offset = current * factor - offset
        factor = Math.floor((value - offset * offset) / factor)
        current = Math.floor((initial + offset) / factor)
        factors.push(current)
    }
    factors.pop()
    return initial
}

function setValue(value, decimal) {
    decimal.value.fill(0)
    decimal.value[0] = value
    decimal.length = 1
}

function getConvergents(previous, current, factors) {
    for (let factor of factors) {
        getSummation(current, factor, previous);
        swapDecimal(previous, current);
    }
}

function getProduct(multiplicand, multiplier, product) {
    for (let i = 0; i < multiplicand.length; ++i) {
        for (let j = 0; j < multiplier.length; ++j) {
            let value = multiplicand.value[i] * multiplier.value[j]
            let index = i + j
            product.value[index] += value % 1e5
            product.value[index + 1] += Math.floor(value / 1e5)
        }
    }
    product.length = multiplicand.length + multiplier.length - 1
    carry(product)
}

function getDouble(decimal) {
    for (let i = 0; i < decimal.length; ++i) {
        decimal.value[i] <<= 1
    }
    carry(decimal)
}

function getSummation(decimal, factor, summation) {
    for (let i = 0; i < decimal.length; ++i) {
        summation.value[i] += decimal.value[i] * factor
    }
    carry(summation)
}

function carry(decimal) {
    for (let i = 0; i < decimal.length; ++i) {
        if (decimal.value[i] >= 1e5) {
            decimal.value[i + 1] += Math.floor(decimal.value[i] / 1e5)
            decimal.value[i] %= 1e5
        }
    }
    if (decimal.value[decimal.length]) {
        ++decimal.length
    }
}

function Decimal(capacity, length) {
    this.value = Array(capacity).fill(0)
    this.length = length
}

module.exports = {
    Problem066 : Problem066
};
